LanguageFeatures/Patterns/constant_A06_t02.dart

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion constantPattern ::= booleanLiteral
+///                   | nullLiteral
+///                   | numericLiteral
+///                   | stringLiteral
+///                   | identifier
+///                   | qualifiedName
+///                   | constObjectExpression
+///                   | 'const' typeArguments? '[' elements? ']'
+///                   | 'const' typeArguments? '{' elements? '}'
+///                   | 'const' '(' expression ')'
+///
+/// A constant pattern determines if the matched value is equal to the
+/// constant's value. We don't allow all expressions here because many
+/// expression forms syntactically overlap other kinds of patterns. We avoid
+/// ambiguity while supporting terse forms of the most common constant
+/// expressions like so:
+/// ...
+/// It is a compile-time error if a constant pattern's value is not a valid
+/// constant expression.
+///
+/// @description Check that it is a compile-time error if a constant pattern's
+/// value is not a valid constant expression. Test expressions which are valid
```

Comment:
Another thing is that we _should_ test the case where 'the constant pattern's value is not a valid constant expression'. 

That would be something like `case a:` where `a` is a non-constant variable, `case const C():` where `C` is a class that has a constructor `C()` but it isn't constant, and `case const (C().foo):` where `C` is a class that has a constant constructor and `foo` is a getter (so everything is fine except that `C().foo` is not a constant expression).

Just a few cases such that we can see that there is a check for "is a constant expression". There's no need to cover expressions that aren't constant in general (because we would have other tests covering that, independently of patterns).

Reply:
> At this point we could also keep it simple and say that this test is concerned with syntax that is a correct <caseHead> in Dart-before-patterns, but is a syntax error in Dart-with-patterns.

That's what I'm doing here. Test description reads "Test expressions which are valid in Dart 2.19 but not valid in Dart-with-patterns".

Reply:
Another thing is that we should test the case where 'the constant pattern's value is not a valid constant expression'.

The previous test `constant_A06_t01' does it

Reply:
OK, I had forgotten that `constant_A06_t01` included `case a:` and such.

But then we should adjust the description here: The current description says

> a constant pattern's value is not a valid constant expression

but that description only fits if we're actually looking at a `<constantPattern>`, and that non-terminal cannot derive `1 + 2`, `1 - 2`, etc.

The description could instead be 'Check that a syntax error is emitted for a term which could be derived from `<caseHead>` in Dart 2.19, but cannot be derived from `<caseHead>` when patterns are added to Dart.'